src/async.ts

```diff
@@ -330,21 +330,27 @@ export class AsyncQueue<T> {
  * logic must not be run in parallel because they use the same resource.
  */
 export class AsyncMutex {
-  private held: boolean;
+  private held: number;
   private awaiters: Array<PromiseAndValueCallback<null>>;
 
   public constructor() {
-    this.held = false;
+    this.held = 0;
     this.awaiters = [];
   }
 
-  public async acquire(): Promise<void> {
```

Comment:
The `async` here wasn't quite right because even though this is meant to be used as an async function, the function body does not use async/await semantics— it constructs Promises explicitly.